Feature/company crud

Yobs2K opened the pull request:
close issue #14

Yobs2K committed:
rename UserModel to SimpleUserModel

Yobs2K committed:
add simple structure for company web-api

Yobs2K committed:
add company crud interface

Yobs2K committed:
implement company crud services

Yobs2K committed:
add company crud controller without @PreAuthorize

Added company controller.
Fixed some logic in services.

Yobs2K committed:
fix CompanyRepository.findAllPageable() method.

Yobs2K committed:
Merge branch 'developer' into feature/company-crud

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyService.java:
```diff
@@ -13,4 +13,16 @@ public interface CompanyService {
+
+    Company addCompany(Company company);
+
+    Company updateCompany(Company company, Long id);
```
update и add один метод, расписал этот момент в тг если что

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +40,54 @@ public Page<Company> getAllPageable(Pageable pageable) {
+    @Override
+    public List<Company> findAllUserCompaniesByUserId(Long userId) {
+        User user = userRepository.findById(userId).orElseThrow(UserNotFoundException::new);
+        return new ArrayList<>(user.getCompanies());
```
не так надо, будем составлять запрос с джоином к UserToCompany и одним запросом все будем получать.

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +40,54 @@ public Page<Company> getAllPageable(Pageable pageable) {
+        oldCompany
+                .setName(newCompany.getName())
+                .setDescription(newCompany.getDescription());
+        return companyRepository.save(oldCompany);
```
по поводу этого метода тоже писал в тг, тут не наадо пока что как либо валидировать

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +40,54 @@ public Page<Company> getAllPageable(Pageable pageable) {
+    @Override
+    public Company updateCompany(Company newCompany, Long id) {
+        Company oldCompany = companyRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("No Company with that ID"));
```
СompanyNotFoundException

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +40,54 @@ public Page<Company> getAllPageable(Pageable pageable) {
+    @Override
+    public Company deleteCompanyById(Long id) {
+        Company deletingCompany = companyRepository.findById(id).orElseThrow(CompanyNotFoundException::new);
+        companyRepository.delete(deletingCompany);
```
deleteByCompanyId

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +40,54 @@ public Page<Company> getAllPageable(Pageable pageable) {
+    }
+
+    @Override
+    public Company addUserToCompanyByIds(Long userId, Long companyId) {
```
Вот связи эти лучше в сервисе UserToCompanyService делать

HardecGoHard commented on db/src/main/java/com/example/db/entity/User.java:
```diff
@@ -64,4 +63,9 @@ public User setBanned(boolean banned) {
         return this;
     }
+
+    User addCompany(Company company) {
```
Лишний метод

Yobs2K replied:
Почему этот метод лишний?

HardecGoHard commented on db/src/main/java/com/example/db/repository/CompanyRepository.java:
```diff
@@ -1,13 +1,21 @@
 
+    Optional<Company> findByNameIgnoreCase(String name);
+
+    Boolean existsByName(String name);
```
хз зачем искат по нейму, ты наверное хотел здесь сделать username?

HardecGoHard commented on db/src/main/java/com/example/db/repository/CompanyRepository.java:
```diff
@@ -1,13 +1,21 @@
+
+    Boolean existsByName(String name);
+
+    @Query("select c from Company c inner join UserToCompany uc where uc.userId = :userId and uc.userRole = :userRole")
```
Тут то что тебе нужно можно с помощью сигнатуры метода делать средствами spring data. А еще давай юзать не  hql, а нейтив квери

HardecGoHard commented on web/src/main/java/com/example/web/api/reference/CompanyController.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("api-web/company")
```
у нас здесь не апи, а сервер сайд рендеринг

HardecGoHard replied:
То есть и ты должен возвращать не модель, а страничку. Смотри приммеры других контроллеров в веб модуле

Yobs2K replied:
Разве в api.reference не именно апи должно быть, а сервер-сайд в controller?

HardecGoHard replied:
ну если у тебя тут референс, то хз зачем ты сюда компании воткнул создание компании и прочие штуки

HardecGoHard commented on web/src/main/java/com/example/web/api/reference/CompanyController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @GetMapping("/belongTo") // Придумать адрес получше
+    public List<CompanyWebModel> getAllAuthUserCompanies(@AuthenticationPrincipal AuthUser authUser) {
+        return companyWebService.findAllAuthUserCompanies(authUser.getUserModel());
```
в метод этот надо не модель пихать, а айди юзера

Yobs2K replied:
Я тут специально сделал модель аутентифицированного юзера, чтобы он мог только свои компании просмотреть, а не для любого другого юзера. Или нужно именно так, чтобы любой для любого мог просмотреть?

HardecGoHard replied:
ну тебе приходит какой то ауф юзер и для него, по его айди, компании выводятся.

HardecGoHard commented on web/src/main/java/com/example/web/api/reference/CompanyController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @GetMapping("/admin") // Придумать адрес получше
+    public List<CompanyWebModel> getAllAuthUserIsAdminCompanies(@AuthenticationPrincipal AuthUser authUser) {
+        return companyWebService.findAllAuthUserIsAdminCompanies(authUser.getUserModel());
```
здесь тоже айди

HardecGoHard commented on web/src/main/java/com/example/web/api/reference/CompanyController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @PostMapping("")
+    public CompanyWebModel createCompany(@AuthenticationPrincipal AuthUser authUser, CompanyWebModel model) {
+        return companyWebService.createCompany(model, authUser.getUserModel());
```
валидацию бы на вебМодель компании, ну и тож айди юзера

HardecGoHard commented on web/src/main/java/com/example/web/api/reference/CompanyController.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @PostMapping("/{companyId}")
+    public CompanyWebModel addUserToCompany(@PathVariable Long companyId, @RequestParam("user_id") Long userId){
```
тут кароч айди юзера лучше передавать в теле запроса

HardecGoHard commented on web/src/main/java/com/example/web/assembler/CompanyWebModelAssembler.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public Company toEntity(CompanyWebModel model) {
```
не, не надо это, если нужно опять энтити получить, то делаешь запрос, так будет лучше. Т.к у тебя в руках будет не энтити. Связей не будет необоходимых и если ты это все попытаешься персистнуть в базу, то все может обосраться

HardecGoHard commented on web/src/main/java/com/example/web/assembler/UserWebModelAssembler.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public User toEntity(UserWebModel model) {
```
ну тож лишнее, см предыдущий коммент

HardecGoHard commented on web/src/main/java/com/example/web/model/CompanyWebModel.java:
```diff
@@ -0,0 +1,26 @@
+package com.example.web.model;
+
+public class CompanyWebModel extends BaseWebModel<CompanyWebModel> {
+
```
нужно прокинуть айди

HardecGoHard commented on web/src/main/java/com/example/web/model/UserWebModel.java:
```diff
@@ -0,0 +1,48 @@
+package com.example.web.model;
+
+public class UserWebModel extends BaseWebModel<UserWebModel> {
+
```
тут тоже айди прокинь

HardecGoHard reviewed (commented):
Так ну комментов оставил, еще много что не написал, Доделаешь это - дополню ещё. Предлагаю в контроллерах принимать в теле не модель, а формы. Например, CreateCompanyForm.

Yobs2K committed:
delete update method in core company service

Yobs2K committed:
Add CreateCompanyForm and ownerId to Company

Added ownerId in Company
Added CreateCompanyForm
Changed web service logic
Deleted unnecessary methods in CompanyRepository
Renamed Company.name to Company.username
Deleted toEntity method in assemblers

Yobs2K committed:
Change argument type in web services

Yobs2K committed:
add UserToCompanyService logic

Yobs2K committed:
fix web assemblers

Yobs2K committed:
delete unnecessary method from CompanyService

Yobs2K committed:
add native query with joins

HardecGoHard commented on core/src/main/java/com/example/core/service/CompanyServiceImpl.java:
```diff
@@ -31,6 +32,28 @@ public Page<Company> getAllPageable(Pageable pageable) {
+
+    @Override
+    public List<Company> findAllUserIsAdminCompaniesByUserId(Long userId) {
+        return companyRepository.findAllUserCompaniesByUserRole(userId, UserCompanyRole.ADMIN);
```
Можно не хардкодить админа, а сделать универсальный метод, который бы принимал роль по которой искать

HardecGoHard commented on core/src/main/java/com/example/core/service/UserToCompanyServiceImpl.java:
```diff
@@ -10,14 +17,34 @@
+        userRepository.save(user.addCompany(company));
+        UserToCompany userToCompany = userToCompanyRepository
+                .findFirstByCompanyIdAndUserId(companyId, userId)
+                .orElseThrow(RuntimeException::new); // Добавить Exception
```
EntityNotFoundException хотя бы

HardecGoHard commented on db/src/main/java/com/example/db/entity/Company.java:
```diff
@@ -1,18 +1,17 @@
 public class Company extends AbstractBaseEntity<Long>{
 
-    private String name;
+    private String username;
```
А нахуя юзернейм в компании

HardecGoHard commented on db/src/main/java/com/example/db/entity/Company.java:
```diff
@@ -1,18 +1,17 @@
+    private String username;
     private String description;
 
+    private Long ownerId;
```
Ну ты овнера добавил, а самого User не добавил

HardecGoHard commented on db/src/main/java/com/example/db/entity/User.java:
```diff
@@ -64,4 +63,9 @@ public User setBanned(boolean banned) {
         return this;
     }
+
+    public User addCompany(Company company) {
```
Не, не надо этот метод тут. Контролировать связями будем на сервисном уровне

Yobs2K committed:
add exceptions and add method to company service

Yobs2K committed:
Merge branch 'developer' into feature/company-crud

Yobs2K committed:
rename property in Company entity

Yobs2K committed:
change CompanyController and add company .ftlh templates

Yobs2K committed:
Add default ftl templates and change pagination

Added simple .ftlh templates for company pages
Pagination - all GET company requests return Page<> now

Yobs2K committed:
rename username property to name in CreateCompanyForm

Yobs2K committed:
fix bug in UserToCompanyServiceImpl

Yobs2K committed:
add creating/changing company forms

HardecGoHard commented on core/src/main/java/com/example/core/exception/EntityNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+        super("Entity not found.");
+    }
+
+    public EntityNotFoundException(String msg) {
```
stat

HardecGoHard commented on core/src/main/java/com/example/core/exception/CompanyNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class CompanyNotFoundException extends EntityNotFoundException {
+    public CompanyNotFoundException() {
+        super("Company entity not found.");
```
test

HardecGoHard commented on core/src/main/java/com/example/core/model/AuthUser.java:
```diff
@@ -6,14 +6,14 @@
-    private final UserModel userModel;
+    private final SimpleUserModel simpleUserModel;
 
     public AuthUser(User user, Collection<? extends GrantedAuthority> authorities) {
```
test
